### ISSUE TITLE
Error presenting challenge: the server could not find the requested resource

### ISSUE THREAD

**Issue by morzan1001** [snowdrop/godaddy-webhook#36]
Hello everyone, 

i have a problem requesting a certificate for a domain of mine. i have made the following configuration:

Certificate:
```
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: <NAME>
  namespace: cert-manager
spec:
  secretName: <NAME>
  renewBefore: 240h
  issuerRef:
    name: letsencrypt-production
    kind: ClusterIssuer
  commonName: "<NAME>"
  dnsNames:
  - "<NAME>"
  - "<NAME>"
```

Issuer:
```
---
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: letsencrypt-production
spec:
  acme:
    server: https://acme-v02.api.letsencrypt.org/directory
    email: <EMAIL>
    privateKeySecretRef:
      name: letsencrypt-production
    solvers:
      - selector:
          dnsZones:
            - "<NAME>"
        dns01:
          webhook:
            config:
              apiKeySecretRef:
                name: godaddy-api-key
                key: token
              production: true
              ttl: 600
            groupName: acme.<NAME>
            solverName: godaddy
```

SecretKey:
```
---
apiVersion: v1
kind: Secret
metadata:
  name: godaddy-api-key
  namespace: cert-manager
type: Opaque
stringData:
  token: <TOKEN>
```

At first i got the error that a resource could not be created for the group i specified. The exact error looked like this:

```
Error presenting challenge: godaddy.acme.<NAME> is forbidden: User "system:serviceaccount:cert-manager:cert-manager" cannot create resource "godaddy" in API group "acme.<NAME>" at the cluster scope 
```

I then created a role-binding to set authorizations correctly if necessary. The whole thing looks like this:
```
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: dns-challenge-missing-role
rules:
- apiGroups: ["acme.<NAME>"]
  resources: ["godaddy"]
  verbs: ["*"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: dns-challenge-missing-role-binding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: dns-challenge-missing-role
subjects:
- kind: ServiceAccount
  name: cert-manager
  namespace: cert-manager
```

Now I get the following error and can't get any further:
```
Error presenting challenge: the server could not find the requested resource (post godaddy.acme.<NAME>) 
```

Maybe someone here has an idea what i have done wrong, i am grateful for any tips :)

**Comment by cmoulliard** [comment]
> kubectl -n cert-manager describe challenge <NAME>

Can you paste the content of `YAML status` please reported by the following command ?
```
kubectl -n cert-manager get challenge <NAME> -oyaml
```

**Comment by morzan1001** [comment]
```
status:
  presented: false
  processing: true
  reason: the server is currently unable to handle the request (post godaddy.acme.mycompany.com)
  state: pending
```

I think this is the same output as in the last output I provided.

**Comment by morzan1001** [comment]
Yes, I think the API itself is not started correctly. The API service has the following status: 
```
status:
  conditions:
    - lastTransitionTime: '2023-12-02T12:50:20Z'
      message: >-
        endpoints for service/godaddy-webhook in "cert-manager" have no
        addresses with port name "https"
      reason: MissingEndpoints
      status: 'False'
      type: Available
```

The pod itself does not seem to start correctly (?) In any case, I cannot view any logs from the pod:
```
kubectl -n cert-manager get pods
```
```
NAME                                     READY   STATUS             RESTARTS         AGE
cert-manager-cainjector-c86f8699-2f2sv   1/1     Running            0                44h
cert-manager-7d9dc8684c-m5kjh            1/1     Running            0                44h
cert-manager-7d9dc8684c-zcwkw            1/1     Running            0                44h
cert-manager-7d9dc8684c-p9f7q            1/1     Running            0                44h
cert-manager-webhook-f8f64cb85-hsjkb     1/1     Running            0                44h
godaddy-webhook-576c45df5f-cwcrl         0/1     CrashLoopBackOff   528 (103s ago)   44h
```

**Comment by cmoulliard** [comment]
Which error do you see (= its log or pod describe) around your `godaddy-webhook-576c45df5f-cwcrl` pod ?

**Comment by cmoulliard** [comment]
> exec /usr/local/bin/webhook: exec format error

Such error is typical about a difference between the compiled go application and platform where it runs - https://stackoverflow.com/questions/50785784/standard-init-linux-go178-exec-user-process-caused-exec-format-error-kuberne. What is the architecture of your k8s cluster (X86, ARM, etc) ?

**Comment by morzan1001** [comment]
The cluster is running on Raspberry PIs -> Architecture is ARM

**Comment by cmoulliard** [comment]
> The cluster is running on Raspberry PIs -> Architecture is ARM

So we know why godaddy webhook fails to start. I will check if I can change the existing [Dockerfile](https://github.com/snowdrop/godaddy-webhook/blob/main/Dockerfile) to release it for ARM platform ;-)

**Comment by morzan1001** [comment]
Thank you very much :+1: 🥇

**Comment by cmoulliard** [comment]
> Raspberry PIs

Something else: Which Pi version do you use and ram capacity you set to run your k8s cluster ?

**Comment by morzan1001** [comment]
I use Raspberry Pi 4 with 8GB ram, but if I look at the workload, Pis with 4 GB ram would probably be enough. So far everything is pretty stable and the installation was easy with K3s.

**Comment by cmoulliard** [comment]
I created a PR to build a multi-arch image - https://github.com/snowdrop/godaddy-webhook/pull/37 @morzan1001

**Comment by morzan1001** [comment]
Things are getting better :D

But unfortunately it doesn't work yet. The container starts now and can also register an API, but unfortunately I am still unhealthy. I can now see the following logs:

```
 kubectl -n cert-manager logs  godaddy-webhook-64cfdcd4f4-d4p4z
```
```
I1204 15:28:40.852556       1 handler.go:232] Adding GroupVersion acme.mycompany.com v1alpha1 to ResourceManager
I1204 15:28:40.864801       1 requestheader_controller.go:169] Starting RequestHeaderAuthRequestController
I1204 15:28:40.864847       1 shared_informer.go:311] Waiting for caches to sync for RequestHeaderAuthRequestController
I1204 15:28:40.864976       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"
I1204 15:28:40.865016       1 shared_informer.go:311] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I1204 15:28:40.865073       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I1204 15:28:40.865092       1 shared_informer.go:311] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I1204 15:28:40.866590       1 dynamic_serving_content.go:132] "Starting controller" name="serving-cert::/tls/tls.crt::/tls/tls.key"
I1204 15:28:40.867887       1 secure_serving.go:210] Serving securely on [::]:443
I1204 15:28:40.867985       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I1204 15:28:40.965192       1 shared_informer.go:318] Caches are synced for RequestHeaderAuthRequestController
I1204 15:28:40.965243       1 shared_informer.go:318] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I1204 15:28:40.965190       1 shared_informer.go:318] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
```
```
kubectl -n cert-manager describe pods
```
```
[...]
Events:
  Type     Reason     Age    From               Message
  ----     ------     ----   ----               -------
  Normal   Scheduled  8m6s   default-scheduler  Successfully assigned cert-manager/godaddy-webhook-64cfdcd4f4-d4p4z to cluster-agent-6
  Normal   Pulling    8m4s   kubelet            Pulling image "quay.io/snowdrop/cert-manager-webhook-godaddy:0.3.0"
  Normal   Pulled     7m57s  kubelet            Successfully pulled image "quay.io/snowdrop/cert-manager-webhook-godaddy:0.3.0" in 7.446531316s (7.446565483s including waiting)
  Normal   Created    7m57s  kubelet            Created container godaddy-webhook
  Normal   Started    7m56s  kubelet            Started container godaddy-webhook
  Warning  Unhealthy  7m55s  kubelet            Readiness probe failed: Get "https://10.42.10.28:443/healthz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
  Warning  Unhealthy  7m54s  kubelet            Readiness probe failed: Get "https://10.42.10.28:443/healthz": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```
```
kubectl get endpoints -n cert-manager      
```
```
NAME                   ENDPOINTS                                         AGE
cert-manager           10.42.5.25:9402,10.42.8.18:9402,10.42.9.14:9402   2d3h
cert-manager-webhook   10.42.3.19:10250                                  2d3h
godaddy-webhook        10.42.10.28:443                                   16m
```

**Comment by cmoulliard** [comment]
Can we then close this ticket as resolve ? @morzan1001

**Comment by morzan1001** [comment]
Yes, I thought you wanted to close the issue together with the pull request. Feel free to close the issue :+1:

**Comment by cmoulliard** [comment]
Fixed with release: 0.3.0
See PR: #37
